test/e2e/e2ethanos/helpers.go

```diff
@@ -19,3 +23,38 @@ func CleanScenario(t *testing.T, s *e2e.Scenario) func() {
 		s.Close()
 	}
 }
+
+func singleJoiningSlash(a, b string) string {
+	aslash := strings.HasSuffix(a, "/")
+	bslash := strings.HasPrefix(b, "/")
+	switch {
+	case aslash && bslash:
+		return a + b[1:]
+	case !aslash && !bslash:
+		return a + "/" + b
+	}
+	return a + b
+}
+
+// NewSingleHostReverseProxy is almost same as httputil.NewSingleHostReverseProxy
+// but it performs a url path rewrite.
+func NewSingleHostReverseProxy(target *url.URL, externalPrefix string) *httputil.ReverseProxy {
+	targetQuery := target.RawQuery
+	director := func(req *http.Request) {
+		req.URL.Scheme = target.Scheme
+		req.URL.Host = target.Host
+		req.URL.Path = singleJoiningSlash(target.Path, strings.TrimPrefix(req.URL.Path, "/"+externalPrefix))
+
+		if targetQuery == "" || req.URL.RawQuery == "" {
+			req.URL.RawQuery = targetQuery + req.URL.RawQuery
+		} else {
+			req.URL.RawQuery = targetQuery + "&" + req.URL.RawQuery
+		}
+
+		if _, ok := req.Header["User-Agent"]; !ok {
+			// Explicitly disable User-Agent so it's not set to default value.
```

Comment:
This doesn't seem to match with what is happening here. We set it to `""` if it does _not_ exist already. What's the rationale behind this?

Reply:
This function is exactly same as https://golang.org/src/net/http/httputil/reverseproxy.go?s=3655:3716#L105
So I didn't thought about it much.

I can't do URL rewrite using `httputil.NewSingleHostReverseProxy` so I just modified it for my needs. The only difference between both of them is that I trim the `externalPrefix` from `url.Path` in my implementation.

Reply:
The commit message here https://github.com/golang/go/commit/6a6c1d9841a1957a2fd292df776ea920ae38ea00 explains my question. Thanks for the pointer!